File: src/main/yummyloop/example/mixin/client/network/OnEntitySpawn.java
package yummyloop.example.mixin.client.network;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.network.ClientPlayNetworkHandler;
import net.minecraft.client.network.packet.EntitySpawnS2CPacket;
import net.minecraft.client.sound.RidingMinecartSoundInstance;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.projectile.ProjectileEntity;
import net.minecraft.entity.vehicle.AbstractMinecartEntity;
import net.minecraft.network.NetworkThreadUtils;
import net.minecraft.network.listener.ClientPlayPacketListener;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
import yummyloop.example.item.spear.SpearEntity;
import yummyloop.example.util.registry.RegistryManager;

@Mixin({ClientPlayNetworkHandler.class})
public abstract class OnEntitySpawn implements ClientPlayPacketListener {
    @Shadow
    private ClientWorld world;
    @Shadow
    private MinecraftClient client;

    @Inject(
            at = @At("HEAD"),
            method = "onEntitySpawn(Lnet/minecraft/client/network/packet/EntitySpawnS2CPacket;)V",
            expect = 0,
            cancellable = true
    )
    private void onOnEntitySpawn(EntitySpawnS2CPacket packet, CallbackInfo info) {
        if (this.client != null && this.world != null) {
            NetworkThreadUtils.forceMainThread(packet, this, this.client);
            EntityType packetEntityType = packet.getEntityTypeId();
            Entity player = world.getEntityById(packet.getEntityData());
            Entity entity = null;
            double x = packet.getX();
            double y = packet.getY();
            double z = packet.getZ();

            EntityType entityType = RegistryManager.INSTANCE.getEntityType().get("spear_entity");
            if (packetEntityType == entityType) {
                entity = new SpearEntity(world, x, y, z);
            }

            if (entity != null) {
                if (player != null && entity instanceof ProjectileEntity) {
                    ((ProjectileEntity) entity).setOwner(player);
                }
                int packetId = packet.getId();
                entity.updateTrackedPosition(x, y, z);
                entity.pitch = (float)(packet.getPitch() * 360) / 256.0F;
                entity.yaw = (float)(packet.getYaw() * 360) / 256.0F;
                entity.setEntityId(packetId);
                entity.setUuid(packet.getUuid());

                world.addEntity(packetId, entity);

                if (entity instanceof AbstractMinecartEntity) {
                    client.getSoundManager().play(new RidingMinecartSoundInstance((AbstractMinecartEntity)entity));
                }
                info.cancel();
            }
        }
    }
}

File: src/main/yummyloop/example/mixin/client/network/OnEntitySpawn.kt
package yummyloop.example.mixin.client.network

import net.minecraft.client.MinecraftClient
import net.minecraft.client.network.ClientPlayNetworkHandler
import net.minecraft.client.network.packet.EntitySpawnS2CPacket
import net.minecraft.client.sound.RidingMinecartSoundInstance
import net.minecraft.client.world.ClientWorld
import net.minecraft.entity.Entity
import net.minecraft.entity.projectile.ProjectileEntity
import net.minecraft.entity.vehicle.AbstractMinecartEntity
import net.minecraft.network.NetworkThreadUtils
import net.minecraft.network.listener.ClientPlayPacketListener
import net.minecraft.util.ThreadExecutor
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import yummyloop.example.ExampleMod
import yummyloop.example.item.Spear

@Mixin(ClientPlayNetworkHandler::class)
abstract class OnEntitySpawn : ClientPlayPacketListener {

    @Shadow private var world: ClientWorld? = null
    @Shadow private var client: MinecraftClient? = null

    @Inject(
            at = [At("HEAD")],
            method = ["Lnet/minecraft/client/network/ClientPlayNetworkHandler;onEntitySpawn(Lnet/minecraft/client/network/packet/EntitySpawnS2CPacket;)V"],
            expect = 0,
            cancellable = true
    )
    fun onOnEntitySpawn(packet : EntitySpawnS2CPacket, info: CallbackInfo) {
        if(client == null || world == null) return

        NetworkThreadUtils.forceMainThread<ClientPlayPacketListener>(packet, this, this.client as ThreadExecutor<*>)
        val x = packet.x
        val y = packet.y
        val z = packet.z
        val entityType = packet.entityTypeId
        val entity: Any?
        val player: Entity?

        when (entityType){
            ExampleMod.spearType -> {
                entity = Spear.SpearEntity(ExampleMod.spearType, world!!, x, y, z)
                player = world!!.getEntityById(packet.entityData)
                if (player != null) {
                    (entity as ProjectileEntity).owner = player
                }
            }
            else -> return
        }

        val packetId = packet.id
        (entity as Entity).updateTrackedPosition(x, y, z)
        (entity as Entity).pitch = (packet.pitch * 360).toFloat() / 256.0f
        (entity as Entity).yaw = (packet.yaw * 360).toFloat() / 256.0f
        (entity as Entity).entityId = packetId
        (entity as Entity).uuid = packet.uuid
        world!!.addEntity(packetId, entity as Entity)
        if (entity is AbstractMinecartEntity) {
            client!!.soundManager.play(RidingMinecartSoundInstance(entity as AbstractMinecartEntity))
        }
        info.cancel()
    }
}
